File: app/store.py
# store.py - 基于本地 JSON 文件的数据存储（免数据库依赖）
import json
import os
import time

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DATA_DIR = os.path.join(BASE_DIR, "data")
DB_FILE = os.path.join(DATA_DIR, "db.json")

DEFAULT_DATA = {
    "cases": [
        {
            "id": "case-0001",
            "name": "示例：百度首页可用性检查",
            "desc": "验证 baidu.com 首页 HTTP 状态码为 200",
            "type": "http",
            "method": "GET",
            "url": "https://www.baidu.com",
            "headers": {},
            "body": "",
            "assertions": [{"type": "status_code", "expected": "200"}],
            "timeout": 10000,
            "enabled": True,
            "createdAt": "",
            "updatedAt": "",
        },
        {
            "id": "case-0002",
            "name": "示例：响应时间低于 3 秒",
            "desc": "检查请求耗时在可接受范围内",
            "type": "http",
            "method": "GET",
            "url": "https://www.qq.com",
            "headers": {},
            "body": "",
            "assertions": [{"type": "response_time", "expected": "3000"}],
            "timeout": 10000,
            "enabled": True,
            "createdAt": "",
            "updatedAt": "",
        },
        {
            "id": "case-0003",
            "name": "示例：example.com 页面标题验证",
            "desc": "用 Playwright 打开 example.com，验证页面标题包含 Example Domain",
            "type": "ui",
            "method": "UI",
            "url": "https://example.com",
            "steps": [
                {"action": "goto", "selector": "", "value": "https://example.com"},
                {"action": "wait_for", "selector": "h1", "value": "10000"},
            ],
            "assertions": [
                {"type": "title_contains", "expected": "Example Domain"},
            ],
            "timeout": 30000,
            "enabled": True,
            "createdAt": "",
            "updatedAt": "",
        },
    ],
    "reports": [],
}


def _ensure_dir():
    os.makedirs(DATA_DIR, exist_ok=True)


def _now():
    return time.strftime("%Y-%m-%dT%H:%M:%S%z", time.localtime())


def read_db():
    _ensure_dir()
    if not os.path.exists(DB_FILE):
        write_db(DEFAULT_DATA)
        return json.loads(json.dumps(DEFAULT_DATA))
    try:
        with open(DB_FILE, "r", encoding="utf-8") as f:
            db = json.load(f)
        db.setdefault("cases", [])
        db.setdefault("reports", [])
        return db
    except (json.JSONDecodeError, OSError):
        write_db(DEFAULT_DATA)
        return json.loads(json.dumps(DEFAULT_DATA))


def write_db(db):
    _ensure_dir()
    tmp = DB_FILE + ".tmp"
    with open(tmp, "w", encoding="utf-8") as f:
        json.dump(db, f, ensure_ascii=False, indent=2)
    os.replace(tmp, DB_FILE)


# ---------- 用例 ----------

def get_cases():
    return read_db()["cases"]


def get_case(case_id):
    return next((c for c in read_db()["cases"] if c["id"] == case_id), None)


def create_case(data: dict) -> dict:
    db = read_db()
    ts = _now()
    new_case = {
        "id": "case-" + str(int(time.time() * 1000))[-8:],
        "name": data.get("name") or "未命名用例",
        "desc": data.get("desc") or "",
        "type": data.get("type") or "http",
        "method": (data.get("method") or "GET").upper(),
        "url": data.get("url") or "",
        "headers": data.get("headers") or {},
        "body": data.get("body") or "",
        "steps": data.get("steps") or [],
        "assertions": data.get("assertions") or [],
        "timeout": int(data.get("timeout") or 10000),
        "enabled": data.get("enabled") is not False,
        "createdAt": ts,
        "updatedAt": ts,
    }
    db["cases"].append(new_case)
    write_db(db)
    return new_case


def update_case(case_id, data: dict) -> dict | None:
    db = read_db()
    for i, c in enumerate(db["cases"]):
        if c["id"] == case_id:
            for key in ("name", "desc", "type", "method", "url", "headers", "body", "steps", "assertions", "timeout", "enabled"):
                if key in data:
                    db["cases"][i][key] = data[key]
            db["cases"][i]["method"] = (db["cases"][i].get("method") or "GET").upper()
            db["cases"][i]["updatedAt"] = _now()
            write_db(db)
            return db["cases"][i]
    return None


def delete_case(case_id) -> bool:
    db = read_db()
    before = len(db["cases"])
    db["cases"] = [c for c in db["cases"] if c["id"] != case_id]
    if len(db["cases"]) == before:
        return False
    write_db(db)
    return True


# ---------- 报告 ----------

def add_report(report: dict) -> dict:
    db = read_db()
    db["reports"].insert(0, report)
    db["reports"] = db["reports"][:100]
    write_db(db)
    return report


def get_reports():
    return read_db()["reports"]


def get_report(report_id):
    return next((r for r in read_db()["reports"] if r["id"] == report_id), None)

File: src/store.js
// store.js - 基于本地 JSON 文件的数据存储（免数据库依赖，开箱即用）
const fs = require('fs');
const path = require('path');

const DATA_DIR = path.join(__dirname, '..', 'data');

const defaultData = {
  cases: [
    {
      id: 'case-0001',
      name: '示例：百度首页可用性检查',
      desc: '验证 baidu.com 首页 HTTP 状态码为 200',
      method: 'GET',
      url: 'https://www.baidu.com',
      headers: {},
      body: '',
      assertions: [{ type: 'status_code', expected: '200' }],
      timeout: 10000,
      enabled: true,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString()
    },
    {
      id: 'case-0002',
      name: '示例：响应时间低于 3 秒',
      desc: '检查请求耗时在可接受范围内',
      method: 'GET',
      url: 'https://www.qq.com',
      headers: {},
      body: '',
      assertions: [{ type: 'response_time', expected: '3000' }],
      timeout: 10000,
      enabled: true,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString()
    }
  ],
  reports: []
};

function ensureDir() {
  if (!fs.existsSync(DATA_DIR)) {
    fs.mkdirSync(DATA_DIR, { recursive: true });
  }
}

function readDB() {
  ensureDir();
  const file = path.join(DATA_DIR, 'db.json');
  if (!fs.existsSync(file)) {
    fs.writeFileSync(file, JSON.stringify(defaultData, null, 2), 'utf-8');
    return JSON.parse(JSON.stringify(defaultData));
  }
  try {
    const raw = fs.readFileSync(file, 'utf-8');
    const parsed = JSON.parse(raw);
    // 保证结构完整
    if (!parsed.cases) parsed.cases = [];
    if (!parsed.reports) parsed.reports = [];
    return parsed;
  } catch (e) {
    console.error('[store] db.json 损坏，重置为默认数据', e.message);
    return JSON.parse(JSON.stringify(defaultData));
  }
}

function writeDB(db) {
  ensureDir();
  const file = path.join(DATA_DIR, 'db.json');
  fs.writeFileSync(file, JSON.stringify(db, null, 2), 'utf-8');
}

function getCases() {
  return readDB().cases;
}

function getCaseById(id) {
  return readDB().cases.find((c) => c.id === id) || null;
}

function createCase(caseData) {
  const db = readDB();
  const now = new Date().toISOString();
  const newCase = {
    id: 'case-' + String(Date.now()).slice(-6) + Math.floor(Math.random() * 90 + 10),
    name: caseData.name || '未命名用例',
    desc: caseData.desc || '',
    method: (caseData.method || 'GET').toUpperCase(),
    url: caseData.url || '',
    headers: caseData.headers || {},
    body: caseData.body || '',
    assertions: Array.isArray(caseData.assertions) ? caseData.assertions : [],
    timeout: Number(caseData.timeout) || 10000,
    enabled: caseData.enabled !== false,
    createdAt: now,
    updatedAt: now
  };
  db.cases.push(newCase);
  writeDB(db);
  return newCase;
}

function updateCase(id, caseData) {
  const db = readDB();
  const idx = db.cases.findIndex((c) => c.id === id);
  if (idx === -1) return null;
  const merged = { ...db.cases[idx], ...caseData, id, updatedAt: new Date().toISOString() };
  merged.method = (merged.method || 'GET').toUpperCase();
  merged.timeout = Number(merged.timeout) || 10000;
  db.cases[idx] = merged;
  writeDB(db);
  return merged;
}

function deleteCase(id) {
  const db = readDB();
  const before = db.cases.length;
  db.cases = db.cases.filter((c) => c.id !== id);
  if (db.cases.length === before) return false;
  writeDB(db);
  return true;
}

function addReport(report) {
  const db = readDB();
  db.reports.unshift(report);
  // 最多保留 100 条历史
  if (db.reports.length > 100) db.reports = db.reports.slice(0, 100);
  writeDB(db);
  return report;
}

function getReports() {
  return readDB().reports;
}

function getReportById(id) {
  return readDB().reports.find((r) => r.id === id) || null;
}

module.exports = {
  getCases,
  getCaseById,
  createCase,
  updateCase,
  deleteCase,
  addReport,
  getReports,
  getReportById
};
